### ISSUE TITLE
Barman can't perform backup (3.9 upgrade issue?)

### ISSUE THREAD

**Issue by sergei-maertens** [EnterpriseDB/barman#856]
Hi - my barman installation appears to not be properly working anymore after I updated packages on my OS.

I received the following cron output causing me to investigate:

```
ERROR: Impossible to start the backup. Check the log for more details, or run 'barman check srv1-regex-it-nl-pg14'
```

My environment:

* Ubuntu 20.04 server (where barman is installed and configured)
* Barman 3.9.0 from the extra repositories

```
root@backups:~# apt-cache show barman
Package: barman
Version: 3.9.0-1.pgdg20.04+1
Architecture: all
Maintainer: Marco Nenciarini <mnencia@debian.org>
Installed-Size: 81
Depends: python3-barman (= 3.9.0-1.pgdg20.04+1), python3-pkg-resources, adduser, rsync, python3:any
Recommends: openssh-server, openssh-client, postgresql-client
Suggests: barman-cli, repmgr
Homepage: https://www.pgbarman.org
Priority: optional
Section: database
Filename: pool/main/b/barman/barman_3.9.0-1.pgdg20.04+1_all.deb
Size: 46212
SHA256: a87592b0ef1d2ff8d262c7568f8f5e270e6830e2f9b5ae3b6c9d9defcf200c38
SHA1: 6506861d28267416578fa5d1f1197a90468d3f5b
MD5sum: 64764e1179b7fad2bbd21727c4742ecd
Description-en: Backup and Recovery Manager for PostgreSQL
 Barman (Backup and Recovery Manager) is an open-source
 administration tool for disaster recovery of PostgreSQL
 servers written in Python.
 .
 It allows your organization to perform remote backups of
 multiple servers in business critical environments to
 reduce risk and help DBAs during the recovery phase.
 .
 Barman is distributed under GNU GPL 3 and maintained
 by 2ndQuadrant.
 .
 This package provides barman binary.
Description-md5: d22bbe67949a3c9d16fae95cbb531954
```

I've updated my system (`apt-get dist-upgrade`) over the weekend and that pulled in the 2-weeks ago released barman 3.9.0.

When I run `barman status all`, the output of the command hangs too:

```
root@backups:~# barman status all
Server geralt-modelbrouwers-nl-pg13:
	Description: PostgreSQL Database (Streaming-Only)
	Active: True
	Disabled: False

```

I had a working setup before, with 3 clusters being managed by barman.

Do you have any pointers of what could be wrong or what else I can investigate?

**Comment by mikewallace1979** [comment]
Hi @sergei-maertens - which Barman version are you upgrading from?

Could also run `barman check srv1-regex-it-nl-pg14` and `barman diagnose` and post the output here, after removing any sensitive information from the output?

Are there any errors in the `barman.log` file?

**Comment by mikewallace1979** [comment]
The timeout while connecting to PostgreSQL does seem to be the most likely reason for the failure.

Barman uses the [psycopg2](https://github.com/psycopg/psycopg2/) library to connect to PostgreSQL which is not used by pg_isready. There is a [report of psycopg2 waiting 127 seconds](https://www.facebook.com/groups/2234775539/posts/10159001593470540/) when attempting an ipv6 connection before falling back to ipv4 - I can't find any reference to this in the psycopg2 github repo, however if this is how psycopg2 behaves then it would explain why the default Barman check timeout of 30 seconds is exceeded in your setup where ipv6 is firewalled and ipv4 isn't.

I haven't been able to verify this report but it might be worth taking a closer look at the ipv6/ipv4 hypothesis - as a test you could try replacing the hostname with the IP4 address in the conninfo string in the Barman config.

**Comment by sergei-maertens** [comment]
Yes, that was indeed my plan. I have to get back to $dayjob now, will report back later. Luckily I'm familiar with psycopg2 so I can dig around too to check if there's a way to force ipv4 if the cause is confirmed.

**Comment by sergei-maertens** [comment]
Confirmed that it works as expected with direct ipv4 IP address as host. I can't find a (documented) way to force ipv4 from a host name, so the resolution will be to either configure the server with the IP address or open up the firewall to accept ipv6 connections.

psycopg2 uses libpq under the hood from what I've gathered, and that doesn't seem to document any such options. Either way, out of scope for barman, I'd say.
